sdk/src/account_context.rs

```diff
@@ -0,0 +1,258 @@
+//! Successors of invoke_context::StackFrame, KeyedAccount and AccountInfo
+
+use crate::{
+    account::{AccountSharedData, ReadableAccount, WritableAccount},
+    instruction::InstructionError,
+    pubkey::Pubkey,
+};
+use std::cell::{RefCell, RefMut};
+
+/// Loaded transaction shared between runtime and programs.
+///
+/// This context is valid for the entire duration of a transaction being processed.
+pub struct TransactionContext {
+    account_keys: Vec<Pubkey>,
+    accounts: Vec<RefCell<AccountSharedData>>,
+    invoke_stack_max_height: usize,
+    invoke_stack: Vec<InstructionContext>,
+    return_data: (Pubkey, Vec<u8>),
+}
+
+impl TransactionContext {
+    pub fn new(
+        keys_and_accounts: Vec<(Pubkey, AccountSharedData)>,
+        invoke_stack_max_height: usize,
+    ) -> Self {
+        let (account_keys, accounts) = keys_and_accounts
+            .into_iter()
+            .map(|(key, account)| (key, RefCell::new(account)))
+            .unzip();
+        Self {
+            account_keys,
+            accounts,
+            invoke_stack_max_height,
+            invoke_stack: Vec::with_capacity(invoke_stack_max_height),
+            return_data: (Pubkey::default(), Vec::new()),
+        }
+    }
+
+    pub fn deconstruct(self) -> Vec<AccountSharedData> {
+        self.accounts
+            .into_iter()
+            .map(|account| account.into_inner())
+            .collect()
+    }
+
+    pub fn number_of_accounts(&self) -> usize {
+        self.accounts.len()
+    }
+
+    pub fn index_of_account(&self, pubkey: &Pubkey) -> Option<usize> {
+        self.account_keys.iter().rev().position(|key| key == pubkey)
+    }
+
+    pub fn invocation_depth(&self) -> usize {
+        self.invoke_stack.len()
+    }
+
+    pub fn invocation_frame_at(
```

Comment:
"invocation_frame" or `InstructionContext`?

Reply:
Renamed all invocation_frame to instruction_context